clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -6183,7 +6486,68 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
   O << "} // namespace detail\n";
   O << "} // namespace _V1\n";
   O << "} // namespace sycl\n";
-  O << "\n";
+
+  FreeFunctionFwdDeclEmitter FFFwdDeclEmitter(O, S.getLangOpts());
+  unsigned ShimCounter = 1;
+  int FreeFunctionCount = 0;
+  for (const KernelDesc &K : KernelDescs) {
+    if (!isFreeFunction(S, K.SyclKernel))
+      continue;
+
+    ++FreeFunctionCount;
+    O << "\n// Definition of " << K.Name << " as a free function kernel\n";
+    if (K.ParamTypes.size() > 0) {
```

Comment:
What is this code doing? Doesn't look like it emits anything in the test

Reply:
There are additions to the int header for free function traits and APIs. That;s what this code does. For example, the int header will now contain entries such as:
```
// Definition of _Z18__sycl_kernel_ff_1Piii as a free function kernel
void ff_1(int *ptr, int start, int end);
static constexpr auto __sycl_shim2() {
  return (void (*)(int *, int, int))ff_1;
}
namespace sycl {
template <>
struct ext::oneapi::experimental::is_kernel<__sycl_shim2()> {
  static constexpr bool value = true;
};
template <>
struct ext::oneapi::experimental::is_nd_range_kernel<__sycl_shim2(), 1> {
  static constexpr bool value = true;
};
}
```

Reply:
I haven't studied the new code but in the previous version, I didn't think these argument visitation was actually emitting anything. The output in the integration header was from emitting the function. What happens to the output in integration header now if you remove lines 6260 - 6264

Reply:
I've removed that loop. It was needed for struct types, but they are not currently supported.